src/analysis/getProjectIds.js

```diff
@@ -0,0 +1,37 @@
+// @flow
+
+// NOTE: This module must be written in vanilla ECMAScript that can be
+// run by Node without a preprocessor. That means that we use `exports`
+// and `require` instead of ECMAScript module keywords, we lazy-load all
+// dependent modules, and we use the Flow comment syntax instead of the
+// inline syntax.
+
+const path = require("path");
+const fs = require("fs");
+
+function getProjectIds(sourcecredDirectory /*: string */) /*: string[] */ {
+  const credPath = path.join(sourcecredDirectory, "data");
+  if (!fs.existsSync(credPath)) {
+    return [];
+  }
+  const filenames = fs.readdirSync(credPath).filter((x) => x.endsWith(".json"));
+  const ids /*: string[] */ = filenames
+    .map((x) => x.slice(0, -5))
+    .map(filenameToId);
+  return ids;
+}
+exports.getProjectIds = getProjectIds;
+
+/**
+ * Convert a project filename back into an id.
+ *
+ * Converts all '$' to '/'. Errors if the filename contains '/'.
```

Comment:
Yeah, take a look at https://github.com/sourcecred/sourcecred/pull/1238 where I've done it using base-64. Is there a particular reason we need base64-url mode?

Could do hashing, but I like having a bijection. Although the way it's implemented in #1238 there's also a `project.json` file stored in every project directory, so we could just use a hash, and then `getProjectIds` can read the project.json file to get the id. It's just a little bit more complexity.

Reply:
+1; this should be base-64. Can’t really be hex-of-UTF-8 because you
can’t assume that the input filenames are UTF-8.

The mirror cache uses base-64.

Reply:
I suggested base64-url because regular base64 contains the `/` in it's character set. Which for filesystems seems like it could cause issues.

About not being able to assume UTF-8. There's no difference there between hex or base64 encoding. You always need to know how to convert from binary to/from the correct characters, regardless of how you encode the binary format. We can assume it however, because we control the filenames and their encoding.

The problem is data corruption / users (or "smart" filesystems) changing filenames. As we will attempt to decode that, it may result in some [very weird strings](https://github.com/minimaxir/big-list-of-naughty-strings), as they would be able to circumvent filesystem protections against illegal characters.

Using hashing removes this entire class of issues, because you'll be forced to work with the assumption you can't decode it.